core/src/main/java/io/parsingdata/metal/data/selection/ByToken.java

```diff
@@ -78,5 +78,22 @@ private static ParseValueList getAllValuesRecursive(final ParseGraph graph, fina
         if (head.isGraph()) { return tailResults.add(getAllValuesRecursive(head.asGraph(), definition)); }
         return tailResults;
     }
+    
+    public static ParseItemList getAllRoots(final ParseGraph graph, final Token definition) {
+        checkNotNull(graph, "graph");
+        checkNotNull(definition, "definition");
+        return (graph.getDefinition() == definition && !graph.isEmpty() ? ParseItemList.create(graph) : ParseItemList.EMPTY)
+            .add(getAllRootsRecursive(graph.head, graph, definition))
+            .add(getAllRootsRecursive(graph.tail, graph, definition));
+    }
+
+    private static ParseItemList getAllRootsRecursive(final ParseItem item, final ParseGraph parent, final Token definition) {
+        final ParseItemList result = (item.getDefinition() == definition && parent.getDefinition() != definition ? ParseItemList.create(item) : ParseItemList.EMPTY);
```

Comment:
This is a pretty long line.

An idea: we introduce a `boolean hasDefinition(Token definition)` on `ParseItem`. This line would become:

```java
final ParseItemList result = (item.hasDefinition(definition) && !parent.hasDefinition(definition) ? ParseItemList.create(item) : ParseItemList.EMPTY);
```

That's still pretty long, though... Nevermind 😉